File: 0086_Partition_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        less, greater = ListNode(), ListNode()
        cur, l, g = head, less, greater
        while cur:
            if cur.val < x:
                l.next = cur
                l = l.next
            else:
                g.next = cur
                g = g.next
            cur = cur.next
        g.next = None
        l.next = greater.next
        return less.next

File: 0086_Partition_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* less = new ListNode();
        ListNode* greater = new ListNode();
        ListNode *curr = head, *cl = less, *cg = greater;
        
        while (curr) {
            if (curr->val < x) {
                cl->next = curr;
                cl = cl->next;
            }
            else {
                cg->next = curr;
                cg = cg->next;
            }
            
            curr = curr->next;
        }
        
        cg->next = NULL;
        cl->next = greater->next;
        return less->next;
    }
};
